tests/func/test_data_storage.py

```diff
@@ -0,0 +1,160 @@
+from typing import Any
+
+import pytest
+
+from datachain.sql.types import (
+    JSON,
+    Array,
+    Boolean,
+    DateTime,
+    Float,
+    Float32,
+    Float64,
+    Int,
+    String,
+)
+from tests.utils import (
+    DEFAULT_TREE,
+    TARRED_TREE,
+    create_tar_dataset_with_legacy_columns,
+)
+
+COMPLEX_TREE: dict[str, Any] = {
+    **TARRED_TREE,
+    **DEFAULT_TREE,
+    "nested": {"dir": {"path": {"abc.txt": "abc"}}},
+}
+
+
+@pytest.mark.parametrize("tree", [COMPLEX_TREE], indirect=True)
+def test_dir_expansion(cloud_test_catalog, version_aware, cloud_type):
+    has_version = version_aware or cloud_type == "gs"
+
+    ctc = cloud_test_catalog
+    session = ctc.session
+    catalog = ctc.catalog
+    src_uri = ctc.src_uri
+    if cloud_type == "file":
+        # we don't want to index things in parent directory
+        src_uri += "/"
+
+    dc = create_tar_dataset_with_legacy_columns(session, ctc.src_uri, "dc")
+    dataset = catalog.get_dataset(dc.name)
+    with catalog.warehouse.clone() as warehouse:
+        dr = warehouse.dataset_rows(dataset, object_name="file")
+        de = dr.dir_expansion()
+        q = de.query(dr.get_table())
+
+        columns = (
+            "id",
+            "is_dir",
+            "source",
+            "path",
+            "version",
+            "location",
+        )
+
+        result = [dict(zip(columns, r)) for r in warehouse.db.execute(q)]
+        to_compare = [(r["path"], r["is_dir"], r["version"] != "") for r in result]
+
+    assert all(r["source"] == ctc.src_uri for r in result)
+
+    # Note, we have both a file and a directory entry for expanded tar files
+    expected = [
+        ("animals.tar", 0, has_version),
+        ("animals.tar", 1, False),
+        ("animals.tar/cats", 1, False),
+        ("animals.tar/cats/cat1", 0, has_version),
+        ("animals.tar/cats/cat2", 0, has_version),
+        ("animals.tar/description", 0, has_version),
+        ("animals.tar/dogs", 1, False),
+        ("animals.tar/dogs/dog1", 0, has_version),
+        ("animals.tar/dogs/dog2", 0, has_version),
+        ("animals.tar/dogs/dog3", 0, has_version),
+        ("animals.tar/dogs/others", 1, False),
+        ("animals.tar/dogs/others/dog4", 0, has_version),
+        ("cats", 1, False),
+        ("cats/cat1", 0, has_version),
+        ("cats/cat2", 0, has_version),
+        ("description", 0, has_version),
+        ("dogs", 1, False),
+        ("dogs/dog1", 0, has_version),
+        ("dogs/dog2", 0, has_version),
+        ("dogs/dog3", 0, has_version),
+        ("dogs/others", 1, False),
+        ("dogs/others/dog4", 0, has_version),
+        ("nested", 1, False),
+        ("nested/dir", 1, False),
+        ("nested/dir/path", 1, False),
+        ("nested/dir/path/abc.txt", 0, has_version),
+    ]
+
+    assert to_compare == expected
+
+
+@pytest.mark.parametrize(
+    "cloud_type,version_aware",
+    [("s3", True)],
+    indirect=True,
+)
+def test_convert_type(cloud_test_catalog):
```

Comment:
This test looks like an unittest. Unrelated, but this probably does not `s3` fixture.